libs/fabric/src/lib/components/group/directives/group-item.directive.ts

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { AfterContentInit, ContentChildren, Directive, Input, OnDestroy, QueryList }
+  from '@angular/core';
+import { IGroup } from 'office-ui-fabric-react';
+import { Subscription } from 'rxjs';
+
+import { ItemChangedPayload } from '../../core/declarative/item-changed';
+import { ChangeableItemDirective } from '../../core/shared/changeable-item.directive';
+
+@Directive({ selector: 'fab-group-item' })
+export class GroupItemDirective extends ChangeableItemDirective<IGroup>
```

Comment:
Seeing that this component is both a `ChangeableItemDirective` as well as a `ChangeableItemsDirective` (being able to contains itself), it makes me think if we should make both `ChangeableItemDirective` and `ChangeableItemDirectives` be their own class, such that you can create an instance of each and then just delegate work to them. The directives can still exist (at the end of the day, in most cases a directive is less likely to be recursive like the case is here), but they'll also just delegate to the standalone classes, acting as a way to save boilerplate in most cases.

pseudocode:
```typescript
// ChangeableItem
export class ChangeableItemHelper<TItem> {
  readonly onItemChanged = new Subject<ItemChangedPayload<string, TItem>>();

  constructor(public readonly key: string) { }

  onChanges(changes: TypedChanges<TItem>) {
    this.onItemChanged.emit({ key: this.key, changes });
  }
}

export abstract class ChangeableItemDirective<TItem>
  implements OnInit, OnChanges<ChangeableItemDirective<TItem>> {

  @Input()
  key: string;

  @Output()
  get onItemChanged(): EventEmitter<ItemChangedPayload<string, TItem>>() {
    return this.changeableItemHelper && this.changeableItemHelper.onItemChanged;
  }

  ngOnInit() {
    this.changeableItemHelper = new ChangeableItemHelper<TItem>(this.key);
  }

  ngOnChanges(changes: TypedChanges<TItem>) {
    this.changeableItemHelper.onChanges(changes);
  }
}

// ChangeableItems
export class ChangeableItemsHelper<TItem> {

  private readonly _subscriptions: Subscription[] = [];

  readonly onItemChanged = new Subject<ChangeableItemDirective<TItem>>>

  constructor(changeableItemHelpers: ReadonlyArray<ChangeableItemHelper<TItem>>) {
    this._subscriptions.push(
      ...changeableItemHelpers.map(changeableItem =>
        changeableItem.onItemChanged.subscribe(changes =>
          this.onItemChanged.emit(changes)
        )
      )
    );
  
	...
  }

  destroy() {
    this._subscriptions.forEach(subscription => subscription.unsubscribe());
  }
}

export abstract class ChangeableItemsDirective<TItem> implements AfterContentInit, OnDestroy {

  @ContentChildren(ChangeableItemDirective)
  readonly directiveItems: QueryList<ChangeableItemDirective<TItem>>;

  @Output()
  get onItemChanged(): EventEmitter<ItemChangedPayload<string, TItem>>() {
    return this.changeableItemsHelper.onItemChanged;
  }

  @Output()
  get onItemsChanged(): EventEmitter<QueryList<ChangeableItemDirective<TItem>>> {
    return this.changeableItemsHelper.onItemsChanged;
  }

  abstract get items(): TItem[];

  ngAfterContentInit() {
    ...
  }
  
  ngOnDestroy() {
    this.changeableItemsHelper.destroy();
  }
}
```

This adds a layer of abstraction, but gives the benefit of re-useable code.

You could also just extract the relevant logic into re-useable, standalone functions and call them in both cases. The thing that mainly bothers me is that I have to choose to inherit from only one of the classes (which makes sense, yet highlights one of the main issues with inheritance over composition).

Reply:
Good point, I was trying to wrap my head around a good way to use properties of both since GroupItemDirective could fall into either category. I think if I add in the logic to remove the self-reference from the list of children directives, this can work for both GroupItem and ContextualMenuItem :)